js/ui/endSessionDialog.js

```diff
@@ -769,6 +843,21 @@ var EndSessionDialog = new Lang.Class({
                 this._type = DialogType.UPGRADE_RESTART;
         }
 
+        this._syncUpdateStatus();
+
+        /* If this._parameters has already saved parameters, it means that we have shown
+         * warning dialog(UPDATES_SCHEDULED DialogType) to the user. So, now skip it and
+         * show regular shutdown dialog.
+         */
+        if (this._hasScheduledUpdates && !this._parameters) {
+            // Show the UPDATES_SCHEDULED dialog before the regular shutdown dialog
+            this._type = DialogType.UPDATES_SCHEDULED;
+
+            // Save the parameters in case the user chooses to "Continue to shutdown"
+            this._parameters = parameters;
+            this._invocation = invocation;
+        }
```

Comment:
Hm, I think this approach is fundamentally flawed. You are injecting `DialogType.UPDATES_SCHEDULED` on top of the regular shutdown type, but the more correct approach would be sending the correct type anyway. It's a laborious path, sure, since it'd involve adding patches to `gnome-session` and hacking other parts of the code; it's also not feasible now, since it'd require refactoring the Automatic Updates management code to an isolated component in GNOME Shell; but this kind of hack makes me nervous about the amount of rework that will have to be done during the rebase.

It's up to you whether you want to chase a more correct approach now, or later, or never, but this trade-off is something we absolutely must keep in mind from now on.

@nedrichards we agreed on having a smaller, less correct patchset, but please take note of this   ↑↑↑